### PR TITLE
CNV consensus (7 of 6): Remove duplicated coordinates and NULLs

### PR DESCRIPTION
**Pull request by fingerfen**
#### Purpose/implementation Section

Remove duplicated coordinates and NULLs in the result file

#### What GitHub issue does your pull request address?

#128 

- [ ] The dependencies required to run the code in this pull request have been added to the project Dockerfile.
- [ ] This analysis has been added to continuous integration.
- [ ] This analysis is recorded in the table in `analyses/README.md`.

### CODE REVIEW

File: analyses/copy_number_consensus_call/Snakefile [hunk 1]
```diff
@@ -189,10 +189,21 @@ rule merge_all:
                           dupdel = ["dup", "del"])
     output:
         ## Define the final output file
-        "results/cnv_consensus.tsv"
+        "../../scratch/endpoints/all_CNVs_combined.tsv"
     shell:
         ## Add a header to the file and combine all of the files using cat. 
-       "echo -e 'chrom\tstart\tend\tmanta_CNVs\tcnvkit_CNVs\tfreec_CNVs\tCNV_type\tBiospecimen\tfile_names' "
+       "echo -e 'chrom\tstart\tend\tmanta_CNVs\tcnvkit_CNVs\tfreec_CNVs\tCNV_type\tsample\tfile_names' "
```
Comment (jashapiro) [inline analyses/copy_number_consensus_call/Snakefile]: Guessing you didn't mean to revert this change.
```suggestion
       "echo -e 'chrom\tstart\tend\tmanta_CNVs\tcnvkit_CNVs\tfreec_CNVs\tCNV_type\ tBiospecimen\tfile_names' "
```

File: analyses/copy_number_consensus_call/src/scripts/remove_dup_NULL_entries.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import argparse
+import sys
+
+# Related third party
+import numpy as np
+import pandas as pd
+
+
+def remove_dup_and_null(cell):
+    """ Remove duplicated coordinates and NULLs """
```
Comment (jashapiro) [inline analyses/copy_number_consensus_call/src/scripts/remove_dup_NULL_entries.py]:
```suggestion
    """ Remove duplicated coordinate sets and NULLs from a comma separated string """
```

File: analyses/copy_number_consensus_call/src/scripts/remove_dup_NULL_entries.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    if "NULL" in set_cell and len(set_cell) > 1:
+        set_cell.remove("NULL")
+
+    ## Join the set
+    new_cell = ",".join(set_cell)
+    return new_cell
+
+
+## Define parser for the input file
+parser = argparse.ArgumentParser(description="""This script removes duplicates coordinates and NULLs from the file""")
```
Comment (jashapiro) [inline analyses/copy_number_consensus_call/src/scripts/remove_dup_NULL_entries.py]:
```suggestion
parser = argparse.ArgumentParser(
    description="""This script removes duplicates coordinate sets and NULLs from the individual callers columns"""
)
```

File: analyses/copy_number_consensus_call/src/scripts/remove_dup_NULL_entries.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+parser.add_argument('--file', required=True,
+                    help='path to the file that needs duplicates and NULLs removed')
+
+args = parser.parse_args()
+
+## Read in the input file
+new_mat = pd.read_csv(args.file, delimiter='\t', header=[0], keep_default_na=False)
+
+## Take out the three columns to be operated on and apply the "remove_dup_and_null" function on those columns
+new_mat[['manta_CNVs','cnvkit_CNVs','freec_CNVs']] = new_mat[['manta_CNVs','cnvkit_CNVs','freec_CNVs']].applymap(lambda x: remove_dup_and_null(x))
```
Comment (jashapiro) [inline analyses/copy_number_consensus_call/src/scripts/remove_dup_NULL_entries.py]: Since your `remove_dup_and_null` function only takes a single argument, you should not need the lambda function definition here.
```suggestion
new_mat[['manta_CNVs','cnvkit_CNVs','freec_CNVs']] = new_mat[['manta_CNVs','cnvkit_CNVs','freec_CNVs']].applymap( remove_dup_and_null)
```

File: analyses/copy_number_consensus_call/src/scripts/remove_dup_NULL_entries.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+    ## Split the cell by commas and convert it to a set, this removes duplicated elements
+    set_cell = set(cell.split(','))
+
+    ## Keep NULLs if it's the only entry in a cell, else, remove NULLs
+    if "NULL" in set_cell and len(set_cell) > 1:
+        set_cell.remove("NULL")
+
+    ## Join the set
+    new_cell = ",".join(set_cell)
```
Comment (jashapiro) [inline analyses/copy_number_consensus_call/src/scripts/remove_dup_NULL_entries.py]: While we are at it, lets make sure these are sorted by the start position:
```suggestion
    sorted_cell = sorted(set_cell, key=lambda x: x.split(":")[0])
    new_cell = ",".join(sorted_cell)
```

**Comment by fingerfen** [comment]
@jashapiro Thanks for catching the reversion. That was not on purpose. Thank you for your effort and quick responses. Also, since we are finished here, I just want to revisit the 3rd (3 of n) PR that I made. Recall you wanted to see if Bedtools Subtract could be a quicker alternative to my Python script. Do you think we should go back at this point to take a look and make changes to that? If I recall correctly, that would be the only part left to improve in this pipeline.

**Comment by jaclyn-taroni** [comment]
@fingerfen - it looks like the last addition before you merged in the master branch (https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/416/commits/7b500c47708fc9683dcc4712bf6ad1da785e0ba8) failed on the copy number consensus step: https://circleci.com/gh/AlexsLemonade/OpenPBTA-analysis/2390

Here's the full output: [build_2390_step_116_container_0.txt](https://github.com/AlexsLemonade/OpenPBTA-analysis/files/4041966/build_2390_step_116_container_0.txt)

**Comment by fingerfen** [comment]
I just tested the line break fix that @jashapiro just did and it worked on my local. I am not sure why it is failing on CI at the momment.

**Comment by jashapiro** [comment]
> I just tested the line break fix that @jashapiro just did and it worked on my local. I am not sure why it is failing on CI at the momment.

It didn't really fail... I just merged in master before it finished running.

**Comment by jashapiro** [comment]
@fingerfen I am just waiting on the update to `cnv_consensus.tsv` to approve this and get it merged in.

**Comment by jashapiro** [comment]
@fingerfen I'm guessing the last commit (f72938a) was to respond to my request for the updated version, but that one doesn't seem to be the latest... it still uses `sample` instead of `Biospecimen`; I'm guessing it is just because you didn't rerun the merge step that sets the header names.

**Comment by fingerfen** [comment]
@jashapiro I apologize. I didn't re-run the merging step. I just re-run it and this new file should have `Biospecimen` instead of `sample`
